docs/rfcs/xxx-pretranspile.md

```diff
@@ -0,0 +1,44 @@
+# Pre-transpiled Atom packages
+
+## Status
+
+Proposed
+
+## Summary
+
+This feature will enable package authors to use conventional npm tooling and package.json conventions to take advantage of JavaScript transpilers like Babel or TypeScript.
+
+## Motivation
+
+Transpiling packages on _publish_ rather than _load_ will have great benefits for package authors:
+
+* Standard `npm` tooling like `prepare` scripts will work for apm packages exactly as they work for npm packages. This will remove the need for custom transpiler pipeline modules like [atom-babel6-transpiler](https://github.com/atom/atom-babel6-transpiler) or [atom-typescript-transpiler](https://github.com/smhxx/atom-ts-transpiler) with their own, independent documentation, configuration and setup.
+* Packages can move transpiler-related dependencies to `devDependencies` and trim installation bloat substantially. (as a data point, the TypeScript compiler is 30MB.)
+* First-time package load will no longer take a hit from transpiling all of the source into the cache. All package loads will benefit from no longer needing to check and side-load from the compiler cache.
+
+## Explanation
+
+### Package publishing
+
+During the `apm publish` call, apm will invoke [`npm pack`](https://docs.npmjs.com/cli/pack) to run all standard npm lifecycle hooks and prepare a `.tar.gz` file. apm then creates a release associated with the created tag and uploads the `.tar.gz` file as an attachment called `package.tar.gz`.
+
+### Package installation
+
+When a user installs a package from atom.io, atom.io first checks to see if its GitHub repository has a release corresponding to that tag with an artifact named `package.tar.gz`. If one is found, the artifact's URL is returned as the `dist` field in the [API response](https://flight-manual.atom.io/atom-server-side-apis/sections/atom-package-server-api/#get-apipackagespackage_nameversionsversion_name). Otherwise, the existing logic is used to return the GitHub tag tarball URL that's returned now.
+
+## Drawbacks
+
+Doing this makes installing a package in production more different than loading it during development. This increases the number of variables that can cause issues between local development and the production of an `apm publish` artifact, like tweaking your `.npmignore` file properly.
```

Comment:
One big problem is native Node modules.  If packages can now ship pre-compiled tarballs, we'd have to include the right native bits for macOS, Linux, and Windows 32-bit and 64-bit.  On top of that, you'll need to have separate builds for each Electron/Node.js ABI version that the package could possibly support, especially as we start upgrading Electron versions more frequently.

The `prebuild` module takes care of this for normal node modules by creating binary tarballs for each platform and ABI version but it quickly spirals out of control: https://github.com/daviwil/node-pty-prebuilt/releases/tag/v0.7.3.  It's also unreasonable to expect the package developer to build these bindings across the 3 supported platforms.

I hesitate to say this, but we might have to provide our own precompilation service which does this on behalf of the package author when they publish an update.  Like, a publish would have to kick off a job on some infrastructure to build the code and handle publishing the tarballs wherever they live.  Not fun...

Reply:
> If packages can now ship pre-compiled tarballs, we'd have to include the right native bits for macOS, Linux, and Windows 32-bit and 64-bit.

The tarball created by `npm pack` does not include anything from transitive dependencies, including binaries:

```
smashwilson @ hubtop ~/src/atom/github (aw/split-github-tab %=) [1]
$ find . -name keytar.node
./node_modules/keytar/build/Release/keytar.node
smashwilson @ hubtop ~/src/atom/github (aw/split-github-tab %=)
$ tar ztvf github-0.18.0-1.tgz | grep 'keytar.node'
smashwilson @ hubtop ~/src/atom/github (aw/split-github-tab %=) [1]
$
```

`apm install` would continue to install those from the appropriate packages in the npm registry, including any that contain native code. Native dependencies would be installed from the appropriate prebuilts or compiled on the user's system, just as they are today.

The only difference would be that apm packages with _top-level_ native dependencies (like, a `binding.gyp` and C++ source in the Atom package itself, not used through an npm dependency) would now be _able_ to distribute prebuilt binaries. Setting that up would be a bit of an adventure, though... personally I'd still prefer to extract the native bits to a different npm package and let other people handle the `prebuilt` setup for me :wink:.

> I hesitate to say this, but we might have to provide our own precompilation service which does this on behalf of the package author when they publish an update.

That would be kind of cool to offer 🤔. I know the experience of, say, Windows users who want to install an Atom package with native dependencies is not ideal - I believe they need to install Node and hunt down native build tools and such, correct?

I think that would be a separate initiative, though.

Reply:
Yeah, that's the biggest problem that Windows users have, they need to install the C++ compiler stuff and Python 2.4 just to get a terminal package (without prebuilds) to install correctly.  What you say sounds fine to me though, if we're still running `apm install` on the package after unpacking the tarball I think it'll be fine.